src/com/ms/silverking/cloud/zookeeper/ZooKeeperExtended.java

```diff
@@ -42,1024 +43,1081 @@
+    } else {
+      Log.warning("No callback for: ", ctx);
     }
-    
-    public void delete(String path) throws KeeperException {
-        try {
-            delete(path, -1);
-        } catch (InterruptedException ie) {
-            throw new RuntimeException("panic");
-        }
+  }
+
+  // ChildrenCallback
+  @Override
+  public void processResult(int rc, String path, Object ctx, List<String> children) {
+    ChildrenCallbackOp callbackOp;
+
+    callbackOp = (ChildrenCallbackOp) callbacks.get((ZKRequestUUID) ctx);
+    if (callbackOp != null) {
+      callbackOp.setComplete(rc, path, children);
+    } else {
+      Log.warning("No callback for: ", ctx);
     }
-    
-    public void waitForCompletion(List<? extends AsyncCallbackOp> ops) throws KeeperException {
-        for (AsyncCallbackOp op : ops) {
-            op.waitForCompletion();
-        }
+  }
+
+  //////////////////////
+  // async op creation
+
+  private void addAsyncCallbackOp(AsyncCallbackOp op) {
+    callbacks.put(op.getRequestUUID(), op);
+  }
+
+  private ACLCallbackOp newACLCallbackOp() {
+    ACLCallbackOp op;
+
+    op = new ACLCallbackOp();
+    addAsyncCallbackOp(op);
+    return op;
+  }
+
+  private Children2CallbackOp newChildren2CallbackOp() {
+    Children2CallbackOp op;
+
+    op = new Children2CallbackOp();
+    addAsyncCallbackOp(op);
+    return op;
+  }
+
+  private ChildrenCallbackOp newChildrenCallbackOp() {
+    ChildrenCallbackOp op;
+
+    op = new ChildrenCallbackOp();
+    addAsyncCallbackOp(op);
+    return op;
+  }
+
+  private DataCallbackOp newDataCallbackOp() {
+    DataCallbackOp op;
+
+    op = new DataCallbackOp();
+    addAsyncCallbackOp(op);
+    return op;
+  }
+
+  private StatCallbackOp newStatCallbackOp() {
+    StatCallbackOp op;
+
+    op = new StatCallbackOp();
+    addAsyncCallbackOp(op);
+    return op;
+  }
+
+  private VoidCallbackOp newVoidCallbackOp() {
+    VoidCallbackOp op;
+
+    op = new VoidCallbackOp();
+    addAsyncCallbackOp(op);
+    return op;
+  }
+
+  private StringCallbackOp newStringCallbackOp() {
+    StringCallbackOp op;
+
+    op = new StringCallbackOp();
+    addAsyncCallbackOp(op);
+    return op;
+  }
+
+  ///////////////
+  // operations
+
+  public void setEphemeralInteger(String path, int data) throws KeeperException {
+    setEphemeral(path, NumConversion.intToBytes(data));
+  }
+
+  public void setEphemeral(String path, byte[] data) throws KeeperException {
+    if (exists(path)) {
+      set(path, data);
+    } else {
+      create(path, data, CreateMode.EPHEMERAL);
     }
-    
-    public VoidCallbackOp deleteAsync(String path) {
-        VoidCallbackOp  op;
-        
-        op = newVoidCallbackOp();
-        delete(path, ANY_VERSION, this, op.getRequestUUID());
-        return op;
+  }
+
+  public Stat set(String path, byte[] data, int version) throws KeeperException {
+    try {
+      return setData(path, data, version);
+    } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)